doc/developer/design/20230531_compute_metrics.md

```diff
@@ -0,0 +1,414 @@
+For example, when there is a leak in controller state it might take a long time before that visibly reflects in the process' memory usage.
+
+The Prometheus/Grafana infrastructure in place at Materialize today is suitable for solving the issues outlined above.
+It enables interested parties to judge the health of our production deployments at a glance and enables high-level diagnosis of issues arising in individual environments and processes.
+To make it useful for Compute we need to do the work of identifying and implementing relevant Prometheus metrics.
+
+# Goals
+[goals]: #goals
+
+Enable individuals from Engineering, Support, and other interested groups to:
+
+* Assess the health of Compute components over all environments.
+* Diagnose misbehaving Compute components, like OOMing or non-responsive clusters.
+* Assess performance trends over time, and especially between releases.
+
+The Compute components in scope are:
+
+* Compute controller
+* Compute replicas
+
+# Non-goals
+[non-goals]: #non-goals
+
+We consider the following goals useful but outside the scope of this design:
+
+* Providing analytics data to inform product decisions.
+* Providing observability for the optimizer. This work is tracked in <https://github.com/MaterializeInc/materialize/issues/17592>.
+
+# Explanation
+[explanation]: #explanation
+
+## Prometheus Metrics
+
+This design proposes to introduce a set of new Prometheus metrics, henceforth referred to as "metrics".
+
+Each metric has a [type][metric-types].
+In Materialize, the metric types in use are:
+
+* **Counter**: A numerical value that can only increase or be reset to zero on process restart.
+* **Gauge**: A numerical value that can increase and decrease arbitrarily.
+* **Histogram**: A collection of bucket counts that represents a cumulative histogram.
+
+A metric consists of a name and a set of labels with values (also referred to as "dimensions").
+Metric and label names can mostly be chosen arbitrarily, but [conventions][metric-naming] for them exist.
+One of these conventions is that metric names should have an application prefix, for Materialize that is `mz_`.
+
+For a given metric, each combination of label values defines a time series.
+For example, a metric `request_count` with a single `request_type` label defines one time series per distinct `request_type` value.
+If you have multiple labels, and/or labels with high cardinalities, the number of time series can quickly grow very large.
+This is especially true for histogram metrics, which expose several per-bucket time series.
+Therefore, it is important to limit the number of labels and, if possible, the number of label values.
+
+[metric-types]: https://prometheus.io/docs/concepts/metric_types/
+[metric-naming]: https://prometheus.io/docs/practices/naming/
+
+## Metrics Collection in the Database
+
+The database has two ways of exposing metrics: directly and through the prometheus-exporter.
+
+### Direct Export
```

Comment:
Yes, this is true for prometheus-exporter metrics as well! I think that's the behavior we need to be able to separate metrics by environment (for controller metrics) or compute worker (for replica metrics).

I haven't thought too much about multi-process replicas. When you have, say, a 2-process replica with 4 workers, you'd get the labels:
* (process-0, worker 0)
* (process-0, worker 1)
* (process-1, worker 2)
* (process-1, worker 3)

So in this case the process tag doesn't contribute any additional time series.

Reply:
hmmm I think maybe I wasn't clear. perhaps an example would help? if you run the following in staging us-east-1:

- `mz_envd_up{namespace="environment-82803c07-9246-4d3d-8516-e080dd332293-0"}`
  - this is a promethus-exporter metric
  - this query returns 1 result with a `pod` label of `promsql-exporter-9d5f88d89-fjsz8`
- `mz_persist_metadata_seconds{namespace="environment-82803c07-9246-4d3d-8516-e080dd332293-0"}`
  - this is a direct export metric
  - this query returns 5 results, one for each process/pod. each of them get a different value for the `pod` label

which means any direct export metrics that are scraped from computed (as opposed to ones from the controller in envd) might end up with more copies than one would naturally expect. not necessarily anything to do, just a heads up in case it wasn't already known

Reply:
Ah, when I said "this is true for prometheus-exporter metrics as well" I was thinking of the per-replica metrics supported by the new prometheus-exporter. If you specify that you want to collect a metric per replica, then the exporter will run the metric query on each replica and expose multiple time series with a `replica_full_name` label. It's true that this is not per-pod but per-replica though.

Thanks for the elaboration and sorry for the confusion. I think we are on the same page here :)